comprobo_final_project/scripts/tasks/tag_task.py

```diff
@@ -0,0 +1,173 @@
+#!usr/bin/env python
+
+
+import time
+import numpy as np
+from ..models.robot import Robot as ModelRobot
+from ..simulator.robot import Robot as SimRobot
+from ..simulator.simulation_visualizer import SimulationVisualizer
+from ..gene_alg_2.genetic_algorithm import GeneticAlgorithm
+from ..visualizations import fitness_vs_run
+
+
+class TagTask(object):
+    """
+    Allow two robots to play tag. One robot chases and one robot runs away.
+    """
+
+    def train(self, robots):
+        """
+        Trains the task to find the most fit organism. Organisms are stored in a
+        log file along with their fitness per generation.
+        """
+
+        log_location = 'logs/log_'+str(int(time.time()))+'.csv'
+        print log_location
+
+        GeneticAlgorithm(
+            log_location=log_location,
+            gen_size=100,
+            num_genes=4,
+            elitism_thresh=0.1,
+            crossover_thresh=0.8,
+            mutation_thresh=0.5,
+            fitness_func=self.get_fitness_func(robots)).train()
+
+
+    def visualizer_test(self, robots, organism):
+        """
+        Use the basic visualizer to see what the robot is doing.
+        """
+
+        simulation_visualizer = SimulationVisualizer(robot, real_world_scale=2)
+        self.run_with_setup(robot, organism)
+
+
+    def get_fitness_func(self, robots):
+        """
+        Provides a fitness function for the genetic alorigthm to optimize for.
+        """
+
+        chasing_robot, running_robot = robots 
+
+        # Using a closure here so we can hold our single robot instance
+        def _chaser_fitness_func(organism):
+            positions = self.run_with_setup(chasing_robot, organism)
+
+            distances = [np.sqrt(position.x**2 + position.y**2) \
+                    for position in positions] # all distances from goal
+            fitness = np.mean(distances) # average distance from goal
+
+            return round(fitness, 5)
+
+        return _fitness_func
+
+        def _runner_fitness_func(organism):
+            positions = self.run_with_setup(running_robot, organism)
+
+            distances = [np.sqrt(position.x**2 + position.y**2) \
```

Comment:
Do these make sense? I'm not sure either, but it seems that you're just tracking the distance between the two robots. Is that sufficient for a tag fitness function? What's the difference between chaser and runner then?